Event/Subscriber/CheckReferenceDataOnRemovalSubscriber.php

```diff
@@ -0,0 +1,65 @@
+<?php
+
+namespace Pim\Bundle\CustomEntityBundle\Event\Subscriber;
+
+use Akeneo\Component\StorageUtils\Event\RemoveEvent;
+use Akeneo\Component\StorageUtils\StorageEvents;
+use Pim\Bundle\CustomEntityBundle\Remover\NonRemovableEntityException;
+use Pim\Bundle\CustomEntityBundle\Repository\AttributeRepository;
+use Pim\Component\Catalog\Query\Filter\Operators;
+use Pim\Component\Catalog\Query\ProductQueryBuilderFactoryInterface;
+use Symfony\Component\EventDispatcher\EventSubscriberInterface;
+
+/**
+ * Checks if a reference data could be remove or not
+ *
+ * @author    Romain Monceau <romain@akeneo.com>
+ * @copyright 2018 Akeneo SAS (http://www.akeneo.com)
+ */
+class CheckReferenceDataOnRemovalSubscriber implements EventSubscriberInterface
+{
+    /** @var AttributeRepository */
+    protected $attributeRepository;
+
+    /** @var ProductQueryBuilderFactoryInterface */
+    protected $pqbFactory;
+
+    /**
+     * @param AttributeRepository $attributeRepository
+     * @param ProductQueryBuilderFactoryInterface $pqbFactory
+     */
+    public function __construct(
+        AttributeRepository $attributeRepository,
+        ProductQueryBuilderFactoryInterface $pqbFactory
+    ) {
+        $this->attributeRepository = $attributeRepository;
+        $this->pqbFactory = $pqbFactory;
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public static function getSubscribedEvents()
+    {
+        return [StorageEvents::PRE_REMOVE => 'checkReferenceDataUsage'];
+    }
+
+    /**
+     * Checks if the reference data is used in a product
+     *
+     * @param RemoveEvent $event
+     */
+    public function checkReferenceDataUsage(RemoveEvent $event)
+    {
+        $referenceData = $event->getSubject();
+
+        $attributes = $this->attributeRepository->findByReferenceData($referenceData);
```

Comment:
Test if `$referenceData` is an instance of `AbstractCustomEntity` first
Also, maybe `$referenceData` is not a very good name for this variable ;)